### ISSUE TITLE
liquidprompt won't start after v2.1.0 update: "bash: read: read error: 0: No such device or address"

### ISSUE THREAD

**Issue by KopfKrieg** [nojhan/liquidprompt#713]
Shell: `bash 5.1.16(1)-release`
Operating system: Fedora 36
```bash
$ uname -a
Linux gwynbleidd 5.17.9-300.fc36.x86_64 #1 SMP PREEMPT Wed May 18 15:08:23 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux
bash-5.1$ cat /proc/version
Linux version 5.17.9-300.fc36.x86_64 (mockbuild@bkernel02.iad2.fedoraproject.org) (gcc (GCC) 12.1.1 20220507 (Red Hat 12.1.1-1), GNU ld version 2.37-27.fc36) #1 SMP PREEMPT Wed May 18 15:08:23 UTC 2022
```
Liquidprompt version: v2.1.0

### Steps to Reproduce

1. Install Fedora 36
2. `git clone git@github.com:nojhan/liquidprompt.git && cd liquidprompt && git checkout v2.1.0`
3. `source liquidprompt`

### Expected Behavior
Liquidprompt should be loaded, and work as always (note: the previous version, v2.0.5., works fine)

### Current Behavior
Liquidprompt won't even load due to the following error:
```
bash: read: read error: 0: No such device or address
```

### Possible Solution
I don't know, sorry ¯\\\_(ツ)_/¯

I don't even know how to debug this, shell scripts aren't exactly a favourite of mine 😅

**Comment by Rycieos** [comment]
Unfortunately that error message isn't very helpful. I'll need some debug output to see what the issue is.

Run the command `set -x` and paste the whole output into pastebin or similar. That turns on debug mode in the shell, to turn it back off you can run `set +x`.

**Comment by Rycieos** [comment]
The error message did not appear in that dump. Does the error only show up when you first load liquidprompt? Try running `set -x`, and then `source liquidprompt`.

**Comment by Rycieos** [comment]
Ok, so it has to be one of the other features recently added.

Can you dump your config options? `declare | grep '^LP_'`.

The new features using `read` are python venv, sysfs battery, sysfs temperature, wifi signal, and container environment. Could you try disabling them and narrowing down which is causing the issue?

- `LP_ENABLE_BATT`
- `LP_ENABLE_TEMP`
- `LP_ENABLE_VIRTUALENV`
- `LP_ENABLE_CONTAINER`
- `LP_ENABLE_WIFI_STRENGTH`

**Comment by KopfKrieg** [comment]
I try to disable those options and hopefully that narrows it down.

Here's the output of `declare | grep '^LP_'`:

```bash
$ declare | grep '^LP_'
LP_BATT=
LP_BATTERY_THRESHOLD=75
LP_BRACKET_CLOSE=']'
LP_BRACKET_OPEN='['
LP_COLORMAP=([0]="" [1]=$'\\[\E[32m\\]' [2]=$'\\[\E[1m\E[32m\\]' [3]=$'\\[\E[33m\\]' [4]=$'\\[\E[1m\E[33m\\]' [5]=$'\\[\E[31m\\]' [6]=$'\\[\E[1m\E[31m\\]' [7]=$'\\[\E[30m\E[41m\\]' [8]=$'\\[\E[1m\E[37m\E[41m\\]' [9]=$'\\[\E[1m\E[33m\E[41m\\]')
LP_COLOR_CHANGES=$'\\[\E[31m\\]'
LP_COLOR_CHARGING_ABOVE=$'\\[\E[32m\\]'
LP_COLOR_CHARGING_UNDER=$'\\[\E[33m\\]'
LP_COLOR_COMMITS=$'\\[\E[33m\\]'
LP_COLOR_COMMITS_BEHIND=$'\\[\E[1m\E[31m\\]'
LP_COLOR_DIFF=$'\\[\E[35m\\]'
LP_COLOR_DIRSTACK=$'\\[\E[1m\E[33m\\]'
LP_COLOR_DISCHARGING_ABOVE=$'\\[\E[33m\\]'
LP_COLOR_DISCHARGING_UNDER=$'\\[\E[31m\\]'
LP_COLOR_ERR=$'\\[\E[35m\\]'
LP_COLOR_HOST=
LP_COLOR_HOST_HASH=$'\\[\E[33m\\]'
LP_COLOR_IN_MULTIPLEXER=$'\\[\E[1m\E[34m\\]'
LP_COLOR_JOB_D=$'\\[\E[33m\\]'
LP_COLOR_JOB_R=$'\\[\E[1m\E[33m\\]'
LP_COLOR_JOB_Z=$'\\[\E[1m\E[33m\\]'
LP_COLOR_MARK=$'\\[\E[1m\\]'
LP_COLOR_MARK_ROOT=$'\\[\E[1m\E[31m\\]'
LP_COLOR_MARK_SUDO=$'\\[\E[1m\E[31m\\]'
LP_COLOR_NOWRITE=$'\\[\E[31m\\]'
LP_COLOR_PATH=$'\\[\E(B\E[m\\]'
LP_COLOR_PATH_LAST_DIR=$'\\[\E(B\E[m\E[1m\\]'
LP_COLOR_PATH_ROOT=$'\\[\E[1m\E[33m\\]'
LP_COLOR_PATH_SEPARATOR=$'\\[\E(B\E[m\E[90m\\]'
LP_COLOR_PATH_SHORTENED=$'\\[\E(B\E[m\E[90m\\]'
LP_COLOR_PATH_VCS_ROOT=$'\\[\E(B\E[m\E[1m\\]'
LP_COLOR_PROXY=$'\\[\E[1m\E[34m\\]'
LP_COLOR_RUNTIME=$'\\[\E[33m\\]'
LP_COLOR_SSH=$'\\[\E[34m\\]'
LP_COLOR_SU=$'\\[\E[1m\E[33m\\]'
LP_COLOR_TELNET=$'\\[\E[30m\E[41m\\]'
LP_COLOR_TIME=$'\\[\E[34m\\]'
LP_COLOR_UP=$'\\[\E[32m\\]'
LP_COLOR_USER=
LP_COLOR_USER_ALT=$'\\[\E[1m\\]'
LP_COLOR_USER_LOGGED=
LP_COLOR_USER_ROOT=$'\\[\E[1m\E[33m\\]'
LP_COLOR_VIRTUALENV=$'\\[\E[36m\\]'
LP_COLOR_WRITE=$'\\[\E[32m\\]'
LP_COLOR_X11_OFF=$'\\[\E[33m\\]'
LP_COLOR_X11_ON=$'\\[\E[32m\\]'
LP_DEBUG_TIME=0
LP_DIRSTACK=
LP_DISABLED_VCS_PATHS=()
LP_ENABLE_BATT=0
LP_ENABLE_BZR=0
LP_ENABLE_COLOR=1
LP_ENABLE_DETACHED_SESSIONS=1
LP_ENABLE_DIRSTACK=0
LP_ENABLE_ERROR=1
LP_ENABLE_FOSSIL=0
LP_ENABLE_FQDN=0
LP_ENABLE_GIT=1
LP_ENABLE_HG=1
LP_ENABLE_JOBS=1
LP_ENABLE_LOAD=1
LP_ENABLE_PERM=1
LP_ENABLE_PROXY=1
LP_ENABLE_RUNTIME=1
LP_ENABLE_RUNTIME_BELL=0
LP_ENABLE_SCLS=1
LP_ENABLE_SCREEN_TITLE=0
LP_ENABLE_SHORTEN_PATH=1
LP_ENABLE_SSH_COLORS=0
LP_ENABLE_SUDO=0
LP_ENABLE_SVN=1
LP_ENABLE_TEMP=0
LP_ENABLE_TIME=0
LP_ENABLE_TITLE=0
LP_ENABLE_VCS_ROOT=0
LP_ENABLE_VIRTUALENV=1
LP_ERR=
LP_HOST=
LP_HOSTNAME_ALWAYS=0
LP_HOST_SYMBOL='\h'
LP_JOBS=
LP_LOAD=
LP_LOAD_CAP=2.0
LP_LOAD_THRESHOLD=0.60
LP_MARK=$'\\$\\[\E(B\E[m\\] '
LP_MARK_ADAPTER=⏚
LP_MARK_BATTERY=⌁
LP_MARK_BRACKET_CLOSE=']'
LP_MARK_BRACKET_OPEN='['
LP_MARK_BZR=⚯
LP_MARK_DEFAULT='\$'
LP_MARK_DIRSTACK=⚞
LP_MARK_DISABLED=⌀
LP_MARK_FOSSIL=⌘
LP_MARK_GIT=±
LP_MARK_HG=☿
LP_MARK_LOAD=⌂
LP_MARK_PERM=:
LP_MARK_PREFIX=' '
LP_MARK_PROXY=↥
LP_MARK_SHORTEN_PATH=' … '
LP_MARK_STASH=+
LP_MARK_SVN=‡
LP_MARK_TEMP=θ
LP_MARK_UNTRACKED='*'
LP_MARK_VCSH='|'
LP_OLD_PROMPT_COMMAND=
LP_OLD_PS1='\s-\v\$ '
LP_OLD_PWD=LP:/home/me
LP_OS=Linux
LP_PATH_CHARACTER_KEEP=3
LP_PATH_KEEP=2
LP_PATH_LENGTH=35
LP_PATH_METHOD=truncate_chars_from_path_left
LP_PATH_VCS_ROOT=1
LP_PERCENTS_ALWAYS=1
LP_PERM=$'\\[\E[32m\\]:\\[\E(B\E[m\\]'
LP_PROXY=
LP_PS1=
LP_PS1_POSTFIX=
LP_PS1_PREFIX=
LP_PWD=$'\\[\E(B\E[m\E[1m\\]~\\[\E(B\E[m\\]'
LP_RUNTIME=
LP_RUNTIME_BELL_THRESHOLD=10
LP_RUNTIME_THRESHOLD=2
LP_SCLS=
LP_TEMP=
LP_TEMP_THRESHOLD=60
LP_TIME=
LP_TIME_ANALOG=0
LP_TITLE_CLOSE=$'\a'
LP_TITLE_OPEN=$'\E]0;'
LP_TTYN=0
LP_USER=$'\\u\\[\E(B\E[m\\]'
LP_USER_ALWAYS=1
LP_VCS=
LP_VENV=
$
```

**Comment by KopfKrieg** [comment]
Woohoo, success \o/

The problem occurs if `LP_ENABLE_TEMP` is activated/set to 1. If I disable it/set to 0, everything works as expected.

**Comment by Rycieos** [comment]
Glad we could track it down.

But that is very weird. It is passing the `-r` test, meaning that the file exists and you have permission to read it. What happens if you `cat` it?

```
cat /sys/class/hwmon/hwmon8/temp2_input
file /sys/class/hwmon/hwmon8/temp2_input
ls -l /sys/class/hwmon/hwmon8/temp2_input
```

Unless that isn't the file we are trying to read. The error `read error: 0: No such device or address` makes me wonder why it is saying "0", like that is the filename. What are the contents of the `_LP_LINUX_TEMPERATURE_FILES[@]` array?
```
printf '%q\n' "${_LP_LINUX_TEMPERATURE_FILES[@]}"
```

**Comment by KopfKrieg** [comment]
```bash
$ cat /sys/class/hwmon/hwmon8/temp2_input
/usr/bin/cat: /sys/class/hwmon/hwmon8/temp2_input: No such device or address

$ file /sys/class/hwmon/hwmon8/temp2_input
/sys/class/hwmon/hwmon8/temp2_input: ERROR: cannot read `/sys/class/hwmon/hwmon8/temp2_input' (No such device or address)

$ ls -l /sys/class/hwmon/hwmon8/temp2_input
-r--r--r--. 1 root root 4096 2022-05-25 18:33 /sys/class/hwmon/hwmon8/temp2_input

$ printf '%q\n' "${_LP_LINUX_TEMPERATURE_FILES[@]}"
/sys/class/hwmon/hwmon10/temp1_input
/sys/class/hwmon/hwmon1/temp1_input
/sys/class/hwmon/hwmon4/temp1_input
/sys/class/hwmon/hwmon4/temp2_input
/sys/class/hwmon/hwmon7/temp1_input
/sys/class/hwmon/hwmon8/temp1_input
/sys/class/hwmon/hwmon8/temp2_input
/sys/class/hwmon/hwmon8/temp3_input
/sys/class/hwmon/hwmon8/temp4_input
/sys/class/hwmon/hwmon8/temp5_input
/sys/class/hwmon/hwmon8/temp6_input
/sys/class/hwmon/hwmon8/temp7_input
/sys/class/hwmon/hwmon8/temp8_input
/sys/class/hwmon/hwmon9/temp1_input
/sys/class/hwmon/hwmon9/temp2_input
/sys/class/hwmon/hwmon9/temp3_input
/sys/class/hwmon/hwmon9/temp4_input
/sys/class/hwmon/hwmon9/temp5_input
/sys/class/hwmon/hwmon\*/device/temp\*_input
/sys/devices/platform/coretemp.0/hwmon/hwmon9/temp1_input
/sys/devices/platform/coretemp.0/hwmon/hwmon9/temp2_input
/sys/devices/platform/coretemp.0/hwmon/hwmon9/temp3_input
/sys/devices/platform/coretemp.0/hwmon/hwmon9/temp4_input
/sys/devices/platform/coretemp.0/hwmon/hwmon9/temp5_input
/sys/class/thermal/thermal_zone0/temp
/sys/class/thermal/thermal_zone1/temp
/sys/class/thermal/thermal_zone2/temp
/sys/class/thermal/thermal_zone3/temp
/sys/class/thermal/thermal_zone4/temp
/sys/class/thermal/thermal_zone5/temp
/sys/class/thermal/thermal_zone6/temp
$
```

**Comment by Rycieos** [comment]
Well that has to be a kernel bug. Every test shows that it exists and you have permission to read it, but trying to read it returns a "does not exist" error.

I'll have to work out a way to handle this, probably can suppress the error message.

**Comment by Rycieos** [comment]
I wrote a commit to fix this on the [bugfix/sysfs-invalid-device](https://github.com/nojhan/liquidprompt/tree/bugfix/sysfs-invalid-device) branch. Please test it and see if it resolves your issue.

If you could also run the `tools/external-tool-tester.sh` script and upload the results, that would help me a lot to prevent a regression on this.

**Comment by KopfKrieg** [comment]
Yep, the fix is working, thank you very much! I also ran `tools/external-tool-tester.sh`, and since the tool said I shouldn't copy/paste the output, but upload it somewhere, the logfile is available here: https://absolem.cc/s/js8rrLafH3CZMkP

**Comment by Rycieos** [comment]
Great. Thanks for the script output, that confirmed that my assumption was correct.

I've released [v2.1.1](https://github.com/nojhan/liquidprompt/releases/tag/v2.1.1) containing this fix.
